Add Azure Monitor alerts for app services

lestermarch opened the pull request:
- Add a basic action group for testing.
- Add an App Service metric alert for HTTP 5xx responses.
- Add an App Service metric alert for HTTP response time.
- Add an App Service activity log alert for when an App Service is stopped.
- Add an App Service activity log alert for when an App Service is deleted.

N.B. The above alert rules only apply to Applications Service apps.

andylangridge commented:
Since we will most likely do this for Appeals too, I think we should move the alerts inside the app-service module maybe?

In the module it might be a good idea to use a flag that only creates them on certain environments perhaps.

lestermarch commented:
@andylangridge 

- Monitoring has been moved to the component folder for each service stack.
- Added monitoring for Appeals app services.
- Alert severity is now based on environment. Code to disable alerts in the dev environment has been added in comments to be enabled later once the alerting requirements have been added/ratified.

andylangridge commented:
> @andylangridge
> 
> * Monitoring has been moved to the component folder for each service stack.
> * Added monitoring for Appeals app services.
> * Alert severity is now based on environment. Code to disable alerts in the dev environment has been added in comments to be enabled later once the alerting requirements have been added/ratified.

Ah,  I mean the node-app-service module rather than the component sorry! At the moment there is a bit of duplication we could remove if it were in there.

lestermarch commented:
> Ah, I mean the node-app-service module rather than the component sorry! At the moment there is a bit of duplication we could remove if it were in there.
> 
> You also don't have to output the app service id, and add extra for_each loops. Which makes things much simpler.

Makes sense! Still value in having `monitoring.tf` files in each component app stack (or at least appeals) as there are components other than App Service which may desirable to monitor (Function Apps) and therefore some differentiation between stacks would be required. Absolutely the basic App Service alerts may be moved into the Node App module. It does end up simpler, as you suggest. I'll work on that.

lestermarch commented:
@andylangridge 

- Moved common App Service monitoring to the Node App module. 
- Updated references to work with the new 3.x azurerm provider resources.

andylangridge commented on app/modules/node-app-service/locals.tf:
```diff
@@ -0,0 +1,7 @@
+locals {
+  severity = {
```
Can we have this as a variable instead of a local? Then we can define it in the respective environment files.

It's good practice to keep your config and your code separate. Can get messy if you have a bunch of environment specific things scattered around the code.

lestermarch replied:
I've removed the `locals.tf` and added severity levels as static values for now as we'll likely revisit these later after defining what types of alert constitute a verbose, low, medium, high, or critical alert: https://github.com/Planning-Inspectorate/infrastructure-environments/pull/49/commits/e7e087e9875a1f7e8117b93796466072eeb4be3c

andylangridge commented on app/components/appeals-app-services/monitoring.tf:
```diff
@@ -0,0 +1,27 @@
+# TODO: Add monitoring for function apps, specific to Appeals
```
Do we need this file? Just not a fan of commented out things hanging around the code unless they're going to be re-enabled exactly as they are. Feel like this one won't be?

lestermarch replied:
Removed in https://github.com/Planning-Inspectorate/infrastructure-environments/pull/49/commits/e7e087e9875a1f7e8117b93796466072eeb4be3c and will add Appeals-specific monitoring in a later PR.

andylangridge commented on app/modules/node-app-service/monitoring.tf:
```diff
@@ -0,0 +1,85 @@
+resource "azurerm_monitor_activity_log_alert" "app_service_delete" {
+  name                = "App Service Deleted - ${reverse(split("/", azurerm_linux_web_app.web_app.id))[0]}"
+  resource_group_name = var.resource_group_name
+  enabled             = var.environment == "dev" ? false : true
```
I'd also avoid doing things like this (similar thing to my comment on the local var). Much better to have a variable called something like "alerting_enabled" so it can be defined in the respective environment variable file.

andylangridge reviewed (commented):
Added some comments around variables. Basically, try to avoid adding conditions on environment within the code itself. It's always better to define a variable for that specific condition so its value can be defined on a per environment basis. That way you keep your config all in one place, and it makes re-factoring etc. much easier.

andylangridge commented on app/components/appeals-app-services/variables.tf:
```diff
@@ -111,6 +116,12 @@ variable "logger_level" {
 
+variable "monitoring_alerts_enabled" {
+  default     = false
+  description = "Indicates whether Azure Monitor alerts are enabled for App Service"
```
Small nitpicky thing, but since I've added a module for the function apps too, maybe we should drop the "for App service" bit.

andylangridge commented on app/modules/node-app-service/variables.tf:
```diff
@@ -76,6 +81,12 @@ variable "location" {
 
+variable "monitoring_alerts_enabled" {
+  default     = false
+  description = "Indicates whether Azure Monitor alerts are enabled for App Service"
```
Here too

andylangridge commented on app/stacks/uk-south/appeals-service/variables.tf:
```diff
@@ -147,6 +152,12 @@ variable "logger_level" {
 
+variable "monitoring_alerts_enabled" {
+  default     = false
+  description = "Indicates whether Azure Monitor alerts are enabled for App Service"
```
Here too

andylangridge commented on app/stacks/uk-south/applications-service/variables.tf:
```diff
@@ -115,6 +120,12 @@ variable "logger_level" {
 
+variable "monitoring_alerts_enabled" {
+  default     = false
+  description = "Indicates whether Azure Monitor alerts are enabled for App Service"
```
Here too

andylangridge commented on app/stacks/uk-west/appeals-service/variables.tf:
```diff
@@ -107,6 +112,12 @@ variable "logger_level" {
 
+variable "monitoring_alerts_enabled" {
+  default     = false
+  description = "Indicates whether Azure Monitor alerts are enabled for App Service"
```
Here too

andylangridge commented on app/stacks/uk-west/applications-service/variables.tf:
```diff
@@ -109,6 +114,12 @@ variable "logger_level" {
 
+variable "monitoring_alerts_enabled" {
+  default     = false
+  description = "Indicates whether Azure Monitor alerts are enabled for App Service"
```
here too
